### PR TITLE
Improve script builtin checking using schema

### PR DESCRIPTION
**Pull request by zdevito**
* This add aten_schema.h which provides a barebones amount of type and
  argument information about each builtin operator
* emitBuiltinCall is updated to use this information rather than
  aten_dispatch to ensure the operator is correct.
* handling of keyword and position arguments now matches python behavior
* There is no longer a requirement that kwargs be constant or that the
  attributes of an op must be entirely constant or non-constant
* compiler now constructs a non-attributed version of the op first and
  then turns it into the constant-attribute version if all attributes
  are constants.
* default arguments for builtins now work
* SugaredValue::call and similar functions now have SourceRange information
  for their arguments so that error reporting is more accurate

Notes:
* This does not try to merge the builtin checking with python arg parser.
  Given that we will eventually have C10 schema which will replace aten_schema,
  we will eventually have a C++ description of the schema and working of that
  description directly will be the easiest form to understand.
* python function calls and script method calls do not support keyword arguments yet.
  When we add this support we should refactor the handling in tryEmitSchema
  that resolves keywords into a common function.

* default arguments work
* keyword arguments to builtins work (still need to extend to calling python and other script methods)
* much better error reporting for incorrect builtins

Lift any constants to attributes on nodes when possible

Some fail tests

some working tests

### CODE REVIEW

File: torch/csrc/jit/script/compiler.cpp [hunk 1]
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
...
+  for(Value* input : new_inputs) {
+    node->addInput(input);
+  }
+  node->copyAttributes(attributes);
+}
+
+
+static std::shared_ptr<SugaredValue> tryEmitSchema(
+  const FunctionSchema& schema,
+  std::stringstream& failure_messages,
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: nit: `ostringstream`

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
...
+    << " but found " << total_inputs << "\n" << loc << "\n";
+    return nullptr;
   }
-  auto op = findTensorOp(n);
-  if(!op) {
-    n->destroy();
-    if(!required)
-      return nullptr;
-    throw ErrorReport(loc) << "unknown builtin op";
+  // fill in position arguments
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: typo: position -> positional

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
...
-    throw ErrorReport(loc) << "produces an unknown number of outputs, so it cannot be used directly from script methods";
+  // fill in named arguments
+  for(const NamedValue& nv : attributes) {
+    auto idx = schema.argumentIndexWithName(nv.name);
+    if(!idx) {
+      err() << "unknown keyword argument '" << nv.name << "'\n" << nv.loc;
+      return nullptr;
+    }
+    if(positional_inputs[*idx]) {
+      err() << "argument '" <<  nv.name << "' previously set \n" << nv.loc;
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: "previously set" might be unclear, how about "specified twice"?

File: torch/csrc/jit/script/compiler.cpp [hunk 4]
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
...
+    if(!v.value->type()->isSubtypeOf(*arg.type)) {
+      err() << "expected a value of type " << arg.type->name() << " for argument '" << arg.name << "' but found "
+            << v.value->type()->name() << "\n"
+            << v.loc;
+      return nullptr;
+    }
+
+    // we only support lists for builtins, where they must be flattened
+    if(arg.type->kind() == TypeKind::ListType) {
+      auto unpacked = graph->insertNode(graph->createTupleUnpack(v.value));
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: TupleUnpack is a bad name now :confuse

File: torch/csrc/jit/script/compiler.cpp [hunk 5]
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
...
-    auto unpacked = graph->insertNode(graph->createTupleUnpack(first));
-    ensureTensors(loc, unpacked->outputs());
-    n->removeInput(0);
-    for(size_t i = 0; i < unpacked->outputs().size(); ++i) {
-      n->insertInput(i, unpacked->outputs().at(i));
+  }
+
+  // we successfully matched this schema, construct the node
+
+  // note: we always construct a purely positional nodes here
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: nit: unnecessary `a`

File: torch/csrc/jit/script/compiler.cpp [hunk 6]
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
...
+  // DO NOT ADD MORE SPECIAL CASES HERE, REFACTOR INTO A FUNCTION IF
+  // NEEDED
+  if(n->kind() == aten::chunk) {
+    auto value = constant_as<int64_t>(flat_inputs[1]);
+    if(!value) {
+      throw ErrorReport(positional_inputs[1]->loc) << "argument 'chunks' must be a constant";
     }
-  } else {
-    ensureTensors(loc, inputs);
+    num_outputs = *value;
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: Actually what's `num_outputs` for chunk?

**Review by apaszke — APPROVED** [review]
It was a long one but LGTM! Some minor questions/suggestions.

File: test/test_jit.py [hunk 7]
```diff
@@ -1656,11 +1656,11 @@ def func(x):
     def test_keyword(self):
         @torch.jit.script
         def func(x):
-            return torch.sum(x, dim=0, keepdim=True)
+            return torch.sum(x, dim=0)
 
         x = torch.rand(10, dtype=torch.float, requires_grad=True)
         y = func(x)
-        y2 = torch.sum(x, dim=0, keepdim=True)
+        y2 = torch.sum(x, dim=0)
```
[the remark this answers is not in the document]
Comment (zdevito) [reply test/test_jit.py]: There is not issue, but previously it was required. I removed it to add a test for default values.

File: tools/jit/gen_jit_dispatch.py [hunk 8]
```diff
@@ -239,16 +238,19 @@ def emit_decl(decl):
         'returns': [{'name': 'result', 'type': 'int64_t', 'dynamic_type': 'int64_t'}],
     } for name in ['sizes', 'strides', 'dim']]
     aten_decls = load_aten_declarations(declarations) + tensor_impl_methods
-    jit_decls = [d for d in aten_decls if is_jit_op(d)]
+    jit_decls = tuple(d for d in aten_decls if is_jit_op(d))
```
[the remark this answers is not in the document]
Comment (zdevito) [reply tools/jit/gen_jit_dispatch.py]: I guess I just changed this for debugging. I will switch it back.

File: tools/jit/templates/aten_schema.cpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  // However, it is more complicated because of this issue and described below
+
+  // literals are stored uniqued and interned in these arrays:
+
+  // string literals
+  const char* names[] = {
+    ${names}
+  };
+
+  // default argument values for all ops, represented as using tensors via tensor_as
```
[the remark this answers is not in the document]
Comment (zdevito) [reply tools/jit/templates/aten_schema.cpp]: yes

File: tools/jit/templates/aten_schema.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+      result.push_back({ names[a[0]], tensors[a[1]], attributes[a[2]], a[3] != 0 });
+    }
+    return result;
+  };
+
+  for(size_t i = 0; i < n_operators; ++i) {
+    auto & op = operators[i];
+    schemas.push_back({names[op[0]], getArgumentList(op[1]), getArgumentList(op[2])});
+  }
+  return schemas;
```
[the remark this answers is not in the document]
Comment (zdevito) [reply tools/jit/templates/aten_schema.cpp]: yeah, something non-linear was happening when the size of the static initializer got big.

File: tools/jit/templates/aten_schema.cpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+std::vector<OperatorSchema> & getOperatorSchemas() {
+  static std::vector<OperatorSchema> schema = createOperatorSchemas();
+  return schema;
+}
+
+static SchemaMap createSchemaMap() {
+  auto schemas = getOperatorSchemas();
```
[the remark this answers is not in the document]
Comment (zdevito) [reply tools/jit/templates/aten_schema.cpp]: That is a small omission that causes a big difference in runtime

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 12]
```diff
@@ -320,7 +323,7 @@ struct GraphFuser {
   // turn consumer node n into a fusion group with just n inside
   // to prepare for fusion and replace uses of n with the new group
   Node * createSingletonFusionGroup(Node * n) {
-    auto group = block->owningGraph()->createFusionGroup(getDevice(n));
+    auto group = block->owningGraph()->createFusionGroup(*getDevice(n));
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/passes/graph_fuser.cpp]: It looks like `.value` throws while `*` doesn't check, so I switched it.

File: torch/csrc/jit/script/compiler.cpp [hunk 13]
```diff
@@ -260,97 +263,272 @@ at::Tensor getAttributeValue(const NamedValue& nv) {
...
+  n->setSourceLocation(std::make_shared<SourceRange>(loc));
+  return g.insertNode(n)->output();
+}
+
+Value* createStack(Graph& g, const SourceRange& loc, at::ArrayRef<Value*> inputs) {
+  // bake in constant propagation for the all-constant case because it is
+  // common to see constant lists like [1, 2] passed to attributes
+  bool all_constant = std::all_of(inputs.begin(), inputs.end(), [&](Value* v) {
+    return v->node()->kind() == prim::Constant;
+  });
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: Can you elaborate? I don't understand the issue yet.

File: torch/csrc/jit/script/compiler.cpp (hunk 4 above)
Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: where is the confusion here?

File: torch/csrc/jit/script/compiler.cpp (hunk 6 above)
Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: It is equal to the number of chunks specified in the chunks=argument. This is why we are checking that this argument is constant here.

File: torch/csrc/jit/script/compiler.h [hunk 14]
```diff
@@ -103,19 +113,19 @@ struct SimpleValue : public SugaredValue {
 };
 
 struct BuiltinFunction : public SugaredValue {
-  BuiltinFunction(const std::string& name, Value* value=nullptr)
-    : name(name), value(value) {}
+  BuiltinFunction(const std::string& name, at::optional<NamedValue> value)
+    : name(name), value(std::move(value)) {}
   std::string name;
-  Value* value;
+  at::optional<NamedValue> value;
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/script/compiler.h]: It is self, yes. Commented.

File: torch/csrc/jit/script/compiler.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/script/compiler.cpp]: I thought the function does something else when I was reading it for the first time. Never mind my comment

File: torch/csrc/jit/script/compiler.cpp (hunk 6 above)
Comment (apaszke) [reply torch/csrc/jit/script/compiler.cpp]: I was curious what's in that field in the schema

File: torch/csrc/jit/script/compiler.cpp (hunk 4 above)
Comment (apaszke) [reply torch/csrc/jit/script/compiler.cpp]: We're using tuple unpack for lists. It's generally the same as statically retrieving all elements, right?

File: torch/csrc/jit/script/module.cpp [hunk 15]
```diff
@@ -20,14 +35,20 @@ std::vector<Value*> Method::emit_call_to(SourceRange loc, Method & callee, Array
         << "' is called recursively involving this call site. Recursive calls are not supported";
   }
   auto fn = callee.graph();
-  ensureSizeMatches(loc, callee.num_inputs(), inputs.size(), "inputs");
-  std::vector<Value*> all_inputs = inputs;
+
+  std::stringstream failure_messages;
+  auto all_inputs = tryMatchSchema(
+    callee.schema ? *callee.schema : defaultSchemaFor(callee),
```
Comment (apaszke) [inline torch/csrc/jit/script/module.cpp]: When could it happen that a script function we're calling misses schema?

**Review by apaszke — CHANGES_REQUESTED** [review]
You still haven't changed `createSchemaMap` in a way that would avoid copying all the schemas. Can you please fix that before merging?

**Review by apaszke — APPROVED** [review]

**Comment by yf225** [comment]
The CI failure is unrelated to the recent sccache change. I tried the build a couple times with `MAX_JOBS=1` and can confirm that it always fails at compiling `aten_schema.cpp`. For next step I will log the output of sccache invocation and see what’s the problem there.

File: torch/csrc/jit/script/module.cpp (hunk 15 above)
Comment (zdevito) [reply torch/csrc/jit/script/module.cpp]: Traces don't have schema, so defaultSchemaFor generates reasonable schema for them.

**Comment by yf225** [comment]
@zdevito After inspecting the compilation log, there seems to be a few issues with the generated `aten_schema.cpp` file: 

torch/csrc/jit/generated/aten_schema.cpp: https://gist.githubusercontent.com/yf225/2500ee8d088fa82b8500e298e1c54c0e/raw/5856db317f6128f5c1d1f79e514b96ea55a2a753/generated_aten_schema.cpp

error log: https://gist.githubusercontent.com/yf225/117507fada93ed74f724adcbade04989/raw/8f602e9f96301729716c26fbf0a8aa09e30d121d/sccache_log.txt

I suspect that fixing the "character constant too long for its type" error for all the `odict_keys` will also fix the other errors.

**Comment by zdevito** [comment]
This looks like it was a python 2 vs 3 thing. I think it is fixed now. Do you have a sense of why the error messages from the compilation were not reported in the original log?

**Comment by yf225** [comment]
@zdevito Previously the sccache server was running as a detached daemon and the compilation failure was not sent to stderr. https://github.com/pytorch/pytorch/pull/7555 should fix it.

**Pull request merged**